cpp/benchmarks/common/generate_benchmark_input.cpp

```diff
@@ -0,0 +1,638 @@
+  std::bernoulli_distribution b_dist;
+
+  random_value_fn(distribution_params<bool> const& desc) : b_dist{desc.probability_true} {}
+  bool operator()(std::mt19937& engine) { return b_dist(engine); }
+};
+
+constexpr size_t bitmask_bits = sizeof(cudf::bitmask_type) * 8;
+
+size_t null_mask_size(cudf::size_type num_rows)
+{
+  return (num_rows + bitmask_bits - 1) / bitmask_bits;
+}
+
+bool get_null_mask_bit(std::vector<cudf::bitmask_type> const& null_mask_data, cudf::size_type row)
+{
+  return null_mask_data[row / bitmask_bits] & (cudf::bitmask_type(1) << row % bitmask_bits);
+}
+
+void reset_null_mask_bit(std::vector<cudf::bitmask_type>& null_mask_data, cudf::size_type row)
+{
+  null_mask_data[row / bitmask_bits] &= ~(cudf::bitmask_type(1) << row % bitmask_bits);
+}
+
+template <typename T>
+void set_element_at(T value,
+                    bool valid,
+                    std::vector<T>& values,
+                    std::vector<cudf::bitmask_type>& null_mask,
+                    cudf::size_type idx)
+{
+  if (valid) {
+    values[idx] = value;
+  } else {
+    reset_null_mask_bit(null_mask, idx);
+  }
+}
+
+// identity mapping, except for bools
+template <typename T, typename Enable = void>
+struct stored_as {
+  using type = T;
+};
+
+// Use `int8_t` for bools because that's how they're stored in columns
+template <typename T>
+struct stored_as<T, typename std::enable_if_t<std::is_same<T, bool>::value>> {
+  using type = int8_t;
+};
+
+/**
+ * @brief Creates a column with random content of type @ref T.
+ *
+ * @param profile Parameters for the random generator
+ * @param engine Pseudo-random engine
+ * @param num_rows Size of the output column
+ *
+ * @tparam T Data type of the output column
+ * @return Column filled with random data
+ */
+template <typename T>
+std::unique_ptr<cudf::column> create_random_column(data_profile const& profile,
+                                                   std::mt19937& engine,
+                                                   cudf::size_type num_rows)
+{
+  // Working around vector<bool> and storing bools as int8_t
+  using stored_Type = typename stored_as<T>::type;
+
+  auto valid_dist = std::bernoulli_distribution{1. - profile.get_null_frequency()};
+  auto value_dist = random_value_fn<T>{profile.get_distribution_params<T>()};
+
+  auto const cardinality = std::min(num_rows, profile.get_cardinality());
+  std::vector<stored_Type> samples(cardinality);
+  std::vector<cudf::bitmask_type> samples_null_mask(null_mask_size(cardinality), ~0);
+  for (cudf::size_type si = 0; si < cardinality; ++si) {
+    set_element_at(
+      (stored_Type)value_dist(engine), valid_dist(engine), samples, samples_null_mask, si);
+  }
+
+  // Distribution for picking elements from the array of samples
+  std::uniform_int_distribution<cudf::size_type> sample_dist{0, cardinality - 1};
+  auto const avg_run_len = profile.get_avg_run_length();
+  std::gamma_distribution<float> run_len_dist(4.f, avg_run_len / 4.f);
```

Comment:
This value controls the shape of the curve as such: https://www.itl.nist.gov/div898/handbook/eda/section3/gif/gampdf4.gif
I chose 4 since it gives low probability of getting 0-1 even with lower average run length (~2-3).
I can add a comment about this, but I don't really have anything to derive the number from.

Reply:
👍

Reply:
Moved into a helper function and added a comment about the choice of alpha.